### ISSUE TITLE
exit code 139 (interrupted by signal 11: SIGSEGV) python

### ISSUE THREAD

**Issue by nikorose87** [opensim-org/opensim-core#1632]
Hi Chris, 
Unfortunately, I could do scaling in the python API because I've got some errors setting the markers of a model. A piece of the code where the error is presented is shown below:

```
scalesetup = motionHenryDir+"/Henryscaledef.xml"
Newmarkers = osim.MarkerSet(model,markerspath)
mystate = model.initSystem()
model.replaceMarkerSet(mystate,Newmarkers)
```
And I got the error of the title. So, I don't know if something in the API is not working well or If I'm doing something wrong. 

Thanks and regards.

**Comment by nikorose87** [comment]
Hi again

I've got another problem that generates the same error. In fact, I'm replicating the 'RunTutorialThree.py' but with a personal case of study in python 2.7. However, I got stuck doing scaling, IK and RRA because I haven't found a function that loads the results to the model, so I think that something internally makes the API break.

Here is the piece of code:
```

model = osim.Model(motionHenryDir+'/Henry-Scaled.osim')
mystate = model.initSystem()
#mystate = model.initSystem()
# Set up the visualizer to show the model and simulation
model.setUseVisualizer(True)

# Create a zero value Vec3
zeroVec3 = osim.ArrayDouble.createVec3(0)
# set gravity 0
#model.setGravity(zeroVec3)
# Create the coordinate storage object from the input .sto file
coordinateSto = osim.Storage(motionHenryDir+'/HenryResults.mot')
#model.setAuthors('Nikolay Prieto PhD')

# Get coordinate set from model, and count the number of coordinates
modelCoorSet = model.getCoordinateSet()
nCoords = modelCoorSet.getSize()

## Inverse Kinematics
iktool = osim.InverseKinematicsTool(iksetupfile)
iktool.setModel(model)
iktool.setOutputMotionFileName(IKmodel)
iktool.run()

# Initialize model
mystate = model.initSystem()
# Add a marker set to the model

#newMarkers = osim.MarkerSet(model, Modelfolder + "/gait2354_Scale_MarkerSet.xml")
#tester = model.replaceMarkerSet()

manager = osim.Manager(model)
manager.setInitialTime(0)
manager.setFinalTime(3.0)
manager.integrate(mystate)

```
Thank you.

**Comment by chrisdembia** [comment]
Hey @nikorose87. Thank you for sharing code that causes this error. I have been able to reproduce the SIGSEGV. We have run into similar issues multiple times in the past, and it is usually a simple fix to how SWIG alters memory management between C++ and Python.

> However, I got stuck doing scaling, IK and RRA because I haven't found a function that loads the results to the model, so I think that something internally makes the API break.

I'm not sure what you mean by "I haven't found the function that loads the results into the model." Could you clarify? I'm trying to figure out if this is a separate issue. Are you saying that you want to load the MOT/STO files produced by those tools into Python? For example, using the `Storage` class?

**Comment by chrisdembia** [comment]
Update: Using the latest OpenSim master bindings, I am unable to reproduce this error. Which commit of OpenSim are you using?

**Comment by chrisdembia** [comment]
@nikorose87 If you re-build OpenSim and find that the issue is gone, please close the issue. Otherwise, we can continue troubleshooting. Thank you.

**Comment by nikorose87** [comment]
Chris I'm using Opensim 4.0 swig 3.0.8, I build this opensim version last year on October, may I rebuild the version again?

On the other hand, I used the class `Storage` but I couldn't assign it to the model. And yes! I want to load a mot file in a model and visualize it. 

Thanks!

**Comment by chrisdembia** [comment]
> Chris I'm using Opensim 4.0 swig 3.0.8, I build this opensim version last year on October, may I rebuild the version again?

Ah great, then simply updating your source code should fix your problem. Rebuild the latest master (e.g., pull from GitHub again, or download the latest source zip: https://github.com/opensim-org/opensim-core/archive/master.zip). Note that 4.0 is not released, so it is not clear what you mean by "4.0."

> On the other hand, I used the class Storage but I couldn't assign it to the model. And yes! I want to load a mot file in a model and visualize it.

The new StatesTrajectory class is designed to facilitate this workflow. This question is not well-suited to a GitHub issue, as it is not a bug report or feature request.

**Comment by nikorose87** [comment]
Hi again,

Unfortunately, I made the build twice and I got the following error doing IK:

`InverseKinematicsTool Failed: Error reading MetaData in file '0008~aa~Walking03.trc'. Unexpected key. Expected = DataRate. Received = dataRate.`

Maybe is something related with the uppercase sensitive class recognition.

**Comment by chrisdembia** [comment]
Yes, you should change `dataRate` to `DataRate`. Please close this issue if the original bug is gone.

**Comment by nikorose87** [comment]
Thank you so much for the patience, but my skills in programming are not enough. Finally, the original piece of code of this problem with scaling was able to run. However, when I update the model with `initSystem()`, It produces the same error.

**Comment by tkuchida** [comment]
@nikorose87 Please open an issue on GitHub if you've found a bug in the software; [the OpenSim Forum](https://simtk.org/plugins/phpBB/indexPhpbb.php?f=91) is the appropriate place to post questions about using the software.

**Comment by nikorose87** [comment]
Dear Sir
I'm so sorry about having done questions related with the software. However, another line in my python code produced the same error. Should I have to open a new issue?

**Comment by tkuchida** [comment]
> another line in my python code produced the same error

Please provide a minimal working example. If it helps, [this test](https://github.com/opensim-org/opensim-core/blob/master/Applications/IK/test/testIK.cpp) demonstrates using the InverseKinematicsTool in C++.

**Comment by chrisdembia** [comment]
> Please provide a minimal working example. If it helps, this test demonstrates using the InverseKinematicsTool in C++.

Or, in this case, a minimal failing example. 😆

**Comment by nikorose87** [comment]
Ok sir, so sorry. The scale and IK tools work fine because it prints the results of them. My code is shown below:
```
import opensim as osim

## Load the 2354 model
Modelfolder = "/home/nikorose/Dropbox/Gait_Analysis_data/Models/Gait2392_Simbody"
model = osim.Model(Modelfolder+"/gait2392_simbody.osim")
#mystate = model.initSystem()
# Set up the visualizer to show the model and simulation
model.setUseVisualizer(True)

#Loading Henry's Motion
motionHenryDir = "/home/nikorose/Dropbox/Gait_Analysis_data/Prueba TTA/Corregido"
HenryGRF = motionHenryDir+"/0008~aa~Walking03_grf.mot"
HenryMotion = motionHenryDir+"/0008~aa~Walking03.trc"
staticHenry = motionHenryDir+"/008~aa~Standing.tdf.trc"
#Loading Model's Motion
modelGRF = Modelfolder+"/subject01_walk1_grf.mot"
ModelMotion = Modelfolder+"/subject01_walk1.trc"
ModelStatic = Modelfolder+"/subject01_static.trc"
markerspath = Modelfolder+"/gait2392_Scale_MarkerSet.xml"
iksetupfile = motionHenryDir+'/HenryIKlinux.xml'
IKmodel = 'HenryAPIIK.mot'



## Loading configured files for scaling, IK and ID
scalesetup = motionHenryDir+"/Henryscaledef.xml"
Newmarkers = osim.MarkerSet(model,markerspath)
mystate = model.initSystem()
model.replaceMarkerSet(mystate,Newmarkers)

# Create a zero value Vec3
zeroVec3 = osim.ArrayDouble.createVec3(0)
# set gravity 0
model.setGravity(zeroVec3)
mystate = model.initSystem()

# Create the coordinate storage object from the input .sto file
#coordinatesSto = osim.Storage(motionModel)
# Define some of the subject measurements
subjectMass		=	72.6

# Get coordinate set from model, and count the number of coordinates
modelCoorSet = model.getCoordinateSet()
nCoords = modelCoorSet.getSize()

# Initialize model
mystate = model.initSystem()

## Scaling
scaletool = osim.ScaleTool(scalesetup)
scaletool.getModelScaler().setOutputModelFileName("HenryAPIScaled.osim")
path2sub = scaletool.getPathToSubject()
scaletool.getModelScaler().processModel(model,path2sub,subjectMass)

mystate = model.initSystem()
## Inverse Kinematics
iktool = osim.InverseKinematicsTool(iksetupfile)
iktool.setModel(model)
iktool.setOutputMotionFileName(IKmodel)
iktool.run()

# Add a marker set to the model

#newMarkers = osim.MarkerSet(model, Modelfolder + "/gait2354_Scale_MarkerSet.xml")
#tester = model.replaceMarkerSet()

manager = osim.Manager(model)
manager.setInitialTime(0)
manager.setFinalTime(3.0)
manager.integrate(mystate)
```
After running IK tool, the SIGSEV error appeared and I don't know why.

Thanks in advice.

**Comment by tkuchida** [comment]
Here is a minimal working example in C++:
```cpp
Model model = Model("arm26.osim");
InverseKinematicsTool iktool = InverseKinematicsTool("arm26_Setup_InverseKinematics.xml");
iktool.setModel(model);
iktool.run();
model.initSystem(); //access violation exception
```
In your case, I suggest reading the model again (into another variable). @aseth1 do we want to support this functionality? I suspect that InverseKinematicsTool was not designed to permit reuse of the model after calling `run()`.

**Comment by nikorose87** [comment]
Hi all! 
I don't understand why do I need to update the model into another variable because the IK process doesn't modify or create models. It just creates motions for associating later into a model. That's why I used the `AnalyzeTool` but again The SIGSEV error appeared, here is the code:

```
import opensim as osim

## Load the 2354 model
motionHenryDir = "/home/nikorose/Dropbox/Gait_Analysis_data/Prueba TTA/Corregido"
HenryGRF = motionHenryDir+"/0008~aa~Walking03_grf.mot"
HenryMotion = motionHenryDir+"/0008~aa~Walking03.trc"
staticHenry = motionHenryDir+"/008~aa~Standing.tdf.trc"
scaleSetup = motionHenryDir+'/Henryscaledef.xml'
iksetupfile = motionHenryDir+'/HenryIKlinux.xml'
IKmodel = 'HenryAPIIK.mot'

model = osim.Model(motionHenryDir+'/Henry-Scaled.osim')
mystate = model.initSystem()
#mystate = model.initSystem()
# Set up the visualizer to show the model and simulation
model.setUseVisualizer(False)

# Create a zero value Vec3
zeroVec3 = osim.ArrayDouble.createVec3(0)
# set gravity 0
model.setGravity(zeroVec3)

# Get coordinate set from model, and count the number of coordinates
modelCoorSet = model.getCoordinateSet()
nCoords = modelCoorSet.getSize()

## Inverse Kinematics
iktool = osim.InverseKinematicsTool(iksetupfile)
iktool.setModel(model)
iktool.setOutputMotionFileName(IKmodel)
iktool.run()

## Appending the IK to the model
coordinateSto = osim.Storage(motionHenryDir+'/'+IKmodel)
analysis = osim.AnalyzeTool()
analysis.setModel(model)
analysis.setInitialTime(2.6)
analysis.setFinalTime(5.5)
analysis.setStatesFromMotion(mystate,coordinateSto,True)
analysis.verifyControlsStates()
analysis.run() 
```
Thanks for all!

**Comment by tkuchida** [comment]
Fixed by #1657

**Comment by tkuchida** [comment]
@nikorose87:

> I don't understand why do I need to update the model into another variable because the IK process doesn't modify or create models.

As stated in issue #1656, "InverseKinematicsTool adds a KinematicsReporter for its internal working but doesn't remove it when done". This issue has been fixed by PR #1657, which has been merged into master branch.
